setuptools/command/build_ext.py

```diff
@@ -3,30 +3,33 @@
 import itertools
 from importlib.machinery import EXTENSION_SUFFIXES
 from importlib.util import cache_from_source as _compiled_file_name
-from typing import Dict, Iterator, List, Tuple
+from typing import TYPE_CHECKING, Dict, Iterator, List, Tuple
 from pathlib import Path
 
-from distutils.command.build_ext import build_ext as _du_build_ext
 from distutils.ccompiler import new_compiler
 from distutils.sysconfig import customize_compiler, get_config_var
 from distutils import log
 
 from setuptools.errors import BaseError
 from setuptools.extension import Extension, Library
 
-try:
-    # Attempt to use Cython for building extensions, if available
-    from Cython.Distutils.build_ext import build_ext as _build_ext
+if TYPE_CHECKING:
+    from setuptools.dist import Distribution
+    from distutils.command.build_ext import build_ext as _build_ext
+else:
+    try:
+        # Attempt to use Cython for building extensions, if available
+        from Cython.Distutils.build_ext import build_ext as _build_ext
 
-    # Additionally, assert that the compiler module will load
-    # also. Ref #1229.
-    __import__('Cython.Compiler.Main')
-except ImportError:
-    _build_ext = _du_build_ext
+        # Additionally, assert that the compiler module will load
+        # also. Ref #1229.
+        __import__('Cython.Compiler.Main')
+    except ImportError:
+        from distutils.command.build_ext import build_ext as _build_ext
```

Comment:
I'm not a fan of the extra nesting. Usually, I see special blocks for `if TYPE_CHECKING`, but usually other imports are unaffected. Is it possible to leave this block unconditional and top-level? If not, I'm going to be reluctant to accept the entanglement this approach requires.

Reply:
Unfortunately not, I asked about a similar situation in pyright's GitHub and received the following answer: https://github.com/microsoft/pyright/issues/6156#issuecomment-1762414590
> This is by design, not a bug.
> 
> The determination about what code to analyze and what code to completely ignore is determined in the "binding" phase of the type checker long before types are evaluated and before the code flow graph is established. As per PEP 484, a type checker must honor a specific set of conditional checks including `sys.platform == <literal>`. You're correctly using this in your code, but you're not including the non-Windows code in an `else` statement. If you want this to work in pyright or other static type checking tools, you'd need to change your code as follows:

Reply:
@Avasam, what if `_build_ext`'s type is defined prior to the `try...except` block? Would that be enough to get rid of the conditional?

```python
_build_ext: Type[_du_build_ext]

try:
    ...
```

Reply:
Reverting this I don't see any added type error. Either this was improved, or it's pyright issues it lead to, in which case that'll show up in #4192 . But for this PR I'll just revert to what it was if the change is unnecessary.